feat: 3129 Allow to swipe between product images on the full screen image view

omkarChend1kar opened the pull request:
### What
- Would be nice to be able to swipe to next/previous product image, on [full screen page](https://user-images.githubusercontent.com/5972966/195283693-b291ff2f-be54-44b9-9acd-68750092b059.png).

### Screenshot/Mockup after :
https://user-images.githubusercontent.com/71365710/201989164-8a1053fc-01da-4134-9a86-54c8813111c7.mp4


### Part of 
- #766

omkarChend1kar committed:
Swipeable view for product images

omkarChend1kar committed:
Swipeable view for product images

omkarChend1kar committed:
Merge branch 'swipeable_product_image_view' of https://github.com/omkarChend1kar/smooth-app into swipeable_product_image_view

omkarChend1kar committed:
Merge branch 'develop' of https://github.com/omkarChend1kar/smooth-app into swipeable_product_image_view

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_viewer.dart:
```diff
@@ -42,7 +42,9 @@ class _ProductImageViewerState extends State<ProductImageViewer> {
     imageData = widget.imageData;
-    imageProvider = NetworkImage(imageData.imageUrl!);
+    imageProvider = NetworkImage(
+      imageData.imageUrl ?? '',
```
That's not possible: if there's nothing to display you shouldn't be there with a non-existing image.

omkarChend1kar replied:
Ohh ,Right I should have first check what is behaviour when images are not present, That was presumptuous of me to consider images will always be present.

monsieurtanuki replied:
@omkarChend1kar Thinking again about it, it's not that simple either, as the `imageUrl` can be refreshed from outside (perhaps not yet, but that's part of #3018 that I'm implementing).

That would mean that if on the gallery you click on an existing image, you go to a swipe page with all images, existing or non existing. On a non existing image, you have to display something specific like the "no picture" logo, with a "new image" button: that's what you should change from your PR code.
And if on the gallery you click directly on a non existing image, you go directly to the "new image" feature (like now).

omkarChend1kar replied:
@monsieurtanuki Actually when imageUrl is without proper path or let say empty string, As I have declared there, It shows that No image icon, That's why I was putting empty string there instead of passing null, Since it was throwing nulll exception, After what you said about this, I thought of filtering list with non null Image paths and then show only uploaded images of whichever category would be displayed, 

But I guess this is what you have just mentioned above right,To show the icon/ text for non-existing images.

![noimagedisplay](https://user-images.githubusercontent.com/71365710/202195038-4e349db4-3756-4319-9f08-3e6eb5e98d7c.jpg)

monsieurtanuki replied:
I don't agree: this is not `NetworkImage`'s job to display crap.
That's our job to display something relevant when we do know there's no image there.
And in the previous screen does display something when there's no image - we have a specific "no image" logo. At least for consistency, we should display the same logo.

And, I think I already mentioned that earlier, this in __NOT__ the same behavior when you click on "edit": either you modify an existing image, or you create a new image. In your case you always try to modify, and that will fail with no source image.

omkarChend1kar replied:
Although, I do have one query, After I put our specified logo image at the place of non-existing image source, It doesn't make to display edit button, Instead of that what if we put Add button there and do whatever we do in previous image gallery place such that it will user to add image there itself, Still not sure how going to implement it but just that for non-existing image add buttons looks better.

monsieurtanuki replied:
@omkarChend1kar That's correct - for the "no image" cases we need:
* a specific "no image" logo, instead of the image
* a specific "add" button label, instead of "edit"
* a specific "add" action, instead of "edit"

monsieurtanuki reviewed (commented):
Hi @omkarChend1kar!

I don't have much time to review tonight but there's something I believe is slightly flawed in your code:
* when you click on an existing image, you display the swipe page
* when you click on a non existing image, you go to the "create a new picture" page
* so far so good
* but on the swipe page, you seem to display all images, regardless of them existing or not
* that means crap pages I'm afraid, like in your video you didn't dare to got to "packaging"
* I believe that you should display only displayable images, and not empty pages for empty images

I understood that just reading your code; maybe there's something I'm missing.

omkarChend1kar committed:
Resolve for formatting issue in file smooth_images_view.dart

omkarChend1kar committed:
Resolve for formatting issue in file smooth_images_view.dart

omkarChend1kar committed:
Merge branch 'swipeable_product_image_view' of https://github.com/omkarChend1kar/smooth-app into swipeable_product_image_view

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_swipeable_view.dart:
```diff
@@ -0,0 +1,67 @@
+
+  @override
+  Widget build(BuildContext context) {
+    WidgetsBinding.instance.addPostFrameCallback((_) {
```
Not sure but perhaps would be better in `initState`.

omkarChend1kar replied:
Yeah, Since I didn't use stateful widget, There was no initstate, And only appBar text that needed to be rebuilt, That's why went with Valuenotifier only

omkarChend1kar replied:
Will convert to stateful, That would be more readable.

monsieurtanuki replied:
> Will convert to stateful, That would be more readable.

Good idea. And you can get rid of `ValueNotifier<int>`.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_swipeable_view.dart:
```diff
@@ -0,0 +1,67 @@
+        controller: controller,
+        itemCount: selectedImages.keys.length,
+        itemBuilder: (BuildContext context, int index) {
+          return ProductImageViewer(
```
I'm not sure why you put the `AppBar` here and removed it from `ProductImageViewer`. Wouldn't it work if the `AppBar` was still in `ProductImageViewer`?

omkarChend1kar replied:
The purpose was to make it look like only images are being swiped, If we keep AppBar there itself in ProductImageViewer it looks like as if entire page is being swiped ,And appbar is being swiped is clearly visible, If that's Okay, I will put it back where it was.

monsieurtanuki replied:
You're right, `AppBar` should be moved from `ProductImageViewer`.

monsieurtanuki reviewed (commented):
Hi @omkarChend1kar!
Still not convinced; please have a look at my comments.

Beyond that, as a user I'm a bit confused if
* from page A you see all the images
* if you click on an image in page A you go to page B
* on page B you can swipe towards all page A's images

I know, that's precisely the purpose of the issue/PR: @teolemon just checking, could you confirm that the behavior matches our needs without adding confusion.

For the record, currently we display only the 4 major images, but by the end of the year we should display all the "other" images (tons of them).

Not related to the code:
* could you please specify in the PR comments that it `Closes #3129`.
* it's not a `chore` (or then everything is a chore) (but that's another story), it's a `feat`

omkarChend1kar commented:
> Hi @omkarChend1kar! Still not convinced; please have a look at my comments.
> 
> Beyond that, as a user I'm a bit confused if
> 
> * from page A you see all the images
> * if you click on an image in page A you go to page B
> * on page B you can swipe towards all page A's images
> 
> I know, that's precisely the purpose of the issue/PR: @teolemon just checking, could you confirm that the behavior matches our needs without adding confusion.
> 
> For the record, currently we display only the 4 major images, but by the end of the year we should display all the "other" images (tons of them).
> 
> Not related to the code:
> 
> * could you please specify in the PR comments that it `Closes #3129`.
> * it's not a `chore` (or then everything is a chore) (but that's another story), it's a `feat`

Noted, Will mention it closes #3129

teolemon commented:
- For me, it only shows selected images (front/ingredients/nutrition/packaging), in a swipable manner, so working as intended.
- Other is just a pool of images, some of which being the source of selected images.
- My mockup shows how we might in the future add action button to modify selected images, one of the possible action being to open another view to pick from "other" images. For a reference implementation of that mockup (which is out of the scope for this), you can refer to https://play.google.com/store/apps/details?id=org.openbeautyfacts.scanner (try clicking a photo on the product page to zoom it). It's also the scope of https://github.com/openfoodfacts/smooth-app/issues/766

monsieurtanuki committed:
Merge branch 'develop' into swipeable_product_image_view

monsieurtanuki commented:
> For me, it only shows selected images (front/ingredients/nutrition/packaging), in a swipable manner, so working as intended.
> Other is just a pool of images, some of which being the source of selected images.

@teolemon Then perhaps we would be better off going directly from the product page to the new swipe page and ignore the pool (especially as for the moment it does not display the "other" images). Nothing blocking, though.

teolemon commented:
the pool is absolutely not in my plan for this. You should just see it as a source for a potential image selector in that case.
We could implement it somewhere else in view/hidden mode (like as a card at the bottom of the product page for people who want to moderate).

monsieurtanuki commented:
> the pool is absolutely not in my plan for this. You should just see it as a source for a potential image selector in that case.
> We could implement it somewhere else in view/hidden mode (like as a card at the bottom of the product page for people who want to moderate).

@teolemon Let me rephrase:
* Currently, there is a page ("A") with the 4 main images top to bottom (and not the "other" images)
* When you click on one image in page "A", you go to a detail page ("B") with only one image displayed big
* The purpose of the current PR is to add a swipe effect, in order to navigate between main images while still on page "B".
* The pool page, with tons of all the images, does not exist yet in Smoothie

My previous comment was about the redundancy between page "A" and its 4 images, and page "B" with a single image but instant access to the other 3. I named page "A" the "pool" because that's as close as we are from a pool in Smoothie, and it did provide access to the "other" images (actually it did not because the other images cannot be accessed sync, and I removed them temporarily some weeks ago).

So, when @omkarChend1kar finishes the current PR, we'll have redundant access to the 4 images.
And we could spare one click if the implementation included removing page A and going directly from a click on an image on the product page, to page B on the right image full size.

teolemon commented:
ok, sorry for the misunderstanding.

omkarChend1kar commented:
> > the pool is absolutely not in my plan for this. You should just see it as a source for a potential image selector in that case.
> > We could implement it somewhere else in view/hidden mode (like as a card at the bottom of the product page for people who want to moderate).
> 
> @teolemon Let me rephrase:
> 
> * Currently, there is a page ("A") with the 4 main images top to bottom (and not the "other" images)
> * When you click on one image in page "A", you go to a detail page ("B") with only one image displayed big
> * The purpose of the current PR is to add a swipe effect, in order to navigate between main images while still on page "B".
> * The pool page, with tons of all the images, does not exist yet in Smoothie
> 
> My previous comment was about the redundancy between page "A" and its 4 images, and page "B" with a single image but instant access to the other 3. I named page "A" the "pool" because that's as close as we are from a pool in Smoothie, and it did provide access to the "other" images (actually it did not because the other images cannot be accessed sync, and I removed them temporarily some weeks ago).
> 
> So, when @omkarChend1kar finishes the current PR, we'll have redundant access to the 4 images. And we could spare one click if the implementation included removing page A and going directly from a click on an image on the product page, to page B on the right image full size.

@monsieurtanuki @teolemon Okay, Then are we going with the approach of removing page A and directly going to Page B ?

monsieurtanuki commented:
> @monsieurtanuki @teolemon Okay, Then are we going with the approach of removing page A and directly going to Page B ?

@omkarChend1kar Let's say it's a possible enhancement for a different issue and PR. For the moment, let's implement a swipe gesture that works correctly with existing and non existing images. If ever we decide to remove page A, all the work on page B will already be done.

omkarChend1kar committed:
Merge branch 'develop' of https://github.com/omkarChend1kar/smooth-app into swipeable_product_image_view

omkarChend1kar committed:
1. Refactoring on previous code
2. Displaying default No-image for Non-existing url
3. Displaying add button for Default No-image
4. Enabled action on Add button to add  new image

omkarChend1kar committed:
Merge branch 'swipeable_product_image_view' of https://github.com/omkarChend1kar/smooth-app into swipeable_product_image_view

omkarChend1kar committed:
Resolved formatting and annotations issue

omkarChend1kar commented:
> > @monsieurtanuki @teolemon Okay, Then are we going with the approach of removing page A and directly going to Page B ?
> 
> @omkarChend1kar Let's say it's a possible enhancement for a different issue and PR. For the moment, let's implement a swipe gesture that works correctly with existing and non existing images. If ever we decide to remove page A, all the work on page B will already be done.

Ohkay 👍

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_swipeable_view.dart:
```diff
@@ -0,0 +1,79 @@
+///Widget to display swipeable product images,
+///Opens product image with [initialProductImageDataIndex] from list of images Typecasted from [selectedImages]
+///
+///Field [selectedImages],[initialProductImageDataIndex],[barcode] cannot be null
```
Of course they can't be null, given their types and the `required`. We don't need that comment.

omkarChend1kar replied:
Ohh, Right this versions with Null safety. Will remove those.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_swipeable_view.dart:
```diff
@@ -0,0 +1,79 @@
+}
+
+class _ProductImageSwipeableViewState extends State<ProductImageSwipeableView> {
+  final ValueNotifier<int> currentImageDataIndex = ValueNotifier<int>(0);
```
No added value using a `ValueNotifier`. Please make the code more readable and just use a `late int` that you init in `initState`, and refresh the widget with `setState`.

omkarChend1kar replied:
The thing was, I tried using setState , But its lagging due to entire rendering of the page, And that lag was quite noticable.

monsieurtanuki replied:
I understand how it could be faster with `ValueNotifier` instead of `setState`, but to be honest I'm surprised that it's so noticeable because there's basically only one extra widget to refresh with `setState`: `ProductImageViewer` (that is probably not well optimized).
Please add an explicit comment about your choice, because later developers may ask the same question: "Why not a `setState`?".
In addition to that,  now that you don't use `setState`, you can convert the widget back to stateless, can't you?

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_swipeable_view.dart:
```diff
@@ -0,0 +1,79 @@
+
+class _ProductImageSwipeableViewState extends State<ProductImageSwipeableView> {
+  final ValueNotifier<int> currentImageDataIndex = ValueNotifier<int>(0);
+  List<MapEntry<ProductImageData, ImageProvider?>>? imageList;
```
`late List<MapEntry<ProductImageData, ImageProvider?>> imageList;` instead: it's never going to be null, you just need time to init it.

monsieurtanuki replied:
I don't think we need that. It took me time to understand the concept of `List<MapEntry>`.

I would say:
* there is a `List` of the 4 main `ImageField`s - that's where the order comes from
* there is a `Map<ProductImageData, ImageProvider?>` - that's where the displayed data comes from

omkarChend1kar replied:
> `late List<MapEntry<ProductImageData, ImageProvider?>> imageList;` instead: it's never going to be null, you just need time to init it.

Okay

omkarChend1kar replied:
> `late List<MapEntry<ProductImageData, ImageProvider?>> imageList;` instead: it's never going to be null, you just need time to init it.

Yeah right.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_swipeable_view.dart:
```diff
@@ -0,0 +1,79 @@
+class _ProductImageSwipeableViewState extends State<ProductImageSwipeableView> {
+  final ValueNotifier<int> currentImageDataIndex = ValueNotifier<int>(0);
+  List<MapEntry<ProductImageData, ImageProvider?>>? imageList;
+  PageController? controller;
```
`late PageController controller;` instead: it's never going to be null, you just need time to init it.

omkarChend1kar replied:
Okay

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_swipeable_view.dart:
```diff
@@ -0,0 +1,79 @@
+    controller = PageController(
+      initialPage: widget.initialProductImageDataIndex,
+    );
+    super.initState();
```
`super.initState()`must be the first call of your `initState`, not the last.

omkarChend1kar replied:
Yeah right.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_swipeable_view.dart:
```diff
@@ -0,0 +1,79 @@
+    currentImageDataIndex.value = widget.initialProductImageDataIndex;
+    imageList = widget.selectedImages.entries.toList();
+    controller = PageController(
+      initialPage: widget.initialProductImageDataIndex,
```
Let's say `currentImageDataIndex.value` now that you populated it.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_swipeable_view.dart:
```diff
@@ -0,0 +1,79 @@
+          return ProductImageViewer(
+            selectedImages: widget.selectedImages,
+            barcode: widget.barcode,
+            imageData: imageList![index].key,
```
Here it's a bit complicated: you give twice the same info with different names (`widget.selectedImages` and `imageList![index]`).
What about `widget.selectedImages` and `imageField`?

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_viewer.dart:
```diff
@@ -52,45 +58,68 @@ class _ProductImageViewerState extends State<ProductImageViewer> {
               ),
-              child: PhotoView(
-                minScale: 0.2,
+              child: SmoothImage(
```
I don't know why you change that.
Just test `isImageUrlAvailable` and display either the "no image" logo, or the image the way we displayed it before with `PhotoView`.

omkarChend1kar replied:
Got it

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_viewer.dart:
```diff
@@ -36,14 +37,19 @@ class _ProductImageViewerState extends State<ProductImageViewer> {
   /// When the image is edited, this is the new image
-  late ImageProvider imageProvider;
-  bool _isEdited = false;
+  ImageProvider? imageProvider;
+  bool isImageUrlAvailable = true;
```
It's a bit confusing to have both fields with opposite default values.
Actually I don't think we need `bool isImageUrlAvailable` at all; it's just the same as `imageProvider != null`.

monsieurtanuki replied:
For the record, private fields are supposed to be private. And start with `_`.

omkarChend1kar replied:
> For the record, private fields are supposed to be private. And start with `_`.

Ohkay

omkarChend1kar replied:
> It's a bit confusing to have both fields with opposite default values. Actually I don't think we need `bool isImageUrlAvailable` at all; it's just the same as `imageProvider != null`.

Point was to make it reusable, But the implementation complicated and quite confusing, Will make the changes, Please check it's readable enough.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_swipeable_view.dart:
```diff
@@ -0,0 +1,79 @@
+///Field [selectedImages],[initialProductImageDataIndex],[barcode] cannot be null
+class ProductImageSwipeableView extends StatefulWidget {
+  const ProductImageSwipeableView({
+    Key? key,
```
You can write `super.key` instead.
Personally I don't believe in `key`, but that's another story.

omkarChend1kar replied:
Okay

monsieurtanuki reviewed (commented):
Hi @omkarChend1kar!
Thank you for your code rewritings! Still some things to fix; please have a look at my comments.

omkarChend1kar committed:
Code refactoring

monsieurtanuki commented:
Hi @omkarChend1kar!
There are conflicts to be resolved.

If I were you I wouldn't put too much effort on this PR for the moment, as long as no decision is made about the collision with #3332 (my personal choice being rather with #3332 than with #3129, but I would like to hear a bit from you guys).

That said, whatever you coded regarding the swipe gesture may be reused for the "other" images, but I don't believe we have a very explicit issue about it.

omkarChend1kar commented:
Mockup after changes suggested :


https://user-images.githubusercontent.com/71365710/203113351-83ed1190-e730-4f06-b619-67668a5d4c30.mp4

monsieurtanuki commented:
Thank you @omkarChend1kar for this video!
The swipe is indeed very smooth.

Still:
* there is new code to pull and there are conflicts to resolve - now the upload is "immediate", which is not the case in your video
* that does not solve the collision with #3332 - what is your opinion on the subject?

omkarChend1kar commented:
> Hi @omkarChend1kar! There are conflicts to be resolved.
> 
> If I were you I wouldn't put too much effort on this PR for the moment, as long as no decision is made about the collision with #3332 (my personal choice being rather with #3332 than with #3129, but I would like to hear a bit from you guys).
> 
> That said, whatever you coded regarding the swipe gesture may be reused for the "other" images, but I don't believe we have a very explicit issue about it.

Ohkay

omkarChend1kar commented:
> Thank you @omkarChend1kar for this video! The swipe is indeed very smooth.
> 
> Still:
> 
> * there is new code to pull and there are conflicts to resolve - now the upload is "immediate", which is not the case in your video
> * that does not solve the collision with [Simplify adding photos by removing 2 of the 4 steps #3332](https://github.com/openfoodfacts/smooth-app/issues/3332) - what is your opinion on the subject?

@monsieurtanuki I am not able decide what would be more important from user's perspective, Whether to see all the images with just a swipe or be able to upload the images with few steps ,It depends on user to user and why exactly he/she is visiting the Products page.

monsieurtanuki commented:
> @monsieurtanuki I am not able decide what would be more important from user's perspective, Whether to see all the images with just a swipe or be able to upload the images with few steps ,It depends on user to user and why exactly he is visiting the Products page.

I think the idea is to make the contributions quicker.

M123-dev commented:
My view on this: 
- I like the swiping, in fact im currenly missing in. Not so often I try to swipe through them
- I don't see how this collides with the addition simplification
- As long as we manage to have swipe working together with @monsieurtanuki's latest instant image upload refactoring I don't see why not to merge this

(Havn't had a look at the code)

monsieurtanuki commented:
> * I don't see how this collides with the addition simplification

It does collide because on the same page you would have the crop tool working _and_ the swipe effect - when you swipe, do you swipe to another image or do you move the crop area?

> * As long as we manage to have swipe working together with @monsieurtanuki's latest instant image upload refactoring

Minor rewriting is necessary because @omkarChend1kar started before my code was merged, but it would work.

M123-dev commented:
> It does collide because on the same page you would have the crop tool working and the swipe effect - when you swipe, do you swipe to another image or do you move the crop area?

The crop area shouldn't be swipable, just the image view. When clicking on the edit button you get to a not swipable page. Just like in the normal phone gallery

monsieurtanuki commented:
> The crop area shouldn't be swipable, just the image view. When clicking on the edit button you get to a not swipable page. Just like in the normal phone gallery

Except that what is suggested in #3332 is a unique landing page when you click on an image in the product page.
There's no "edit" button: this page would let you directly change the crop area, take a new photo, confirm and cancel. We get rid of "read only" image pages (where swiping would make sense).
Or - depending on the comment - from product page we land on a page that displays the main 4 images. And then, clicking will lead to the edit page I was referring to - therefore no place for swiping.

omkarChend1kar commented:
> > The crop area shouldn't be swipable, just the image view. When clicking on the edit button you get to a not swipable page. Just like in the normal phone gallery
> 
> Except that what is suggested in #3332 is a unique landing page when you click on an image in the product page. There's no "edit" button: this page would let you directly change the crop area, take a new photo, confirm and cancel. We get rid of "read only" image pages (where swiping would make sense). Or - depending on the comment - from product page we land on a page that displays the main 4 images. And then, clicking will lead to the edit page I was referring to - therefore no place for swiping.

What if we manage to, Enable crop and click new image functionalities with click of edit icon or button just like now ,On current page itself, Without navigating to another screen ,So with click of Edit button swiping will be disabled and If we put a cross/close icon on right corner,It will disable all controls to crop/click new image and enable swiping back. 

Though, Haven't considered how we are going to implement it ,But would it serve the purpose, If we ever manage to do it.

monsieurtanuki commented:
> What if we manage to, Enable crop and click new image functionalities with click of edit icon or button just like now

That would be in contradiction with #3332, wouldn't it?

> On current page itself, Without navigating to another screen

The problem is not "opening a new page". The problem is "having to click on a button" for that.
The idea is to reduce the number of steps / clicks.

> So with click of Edit button swiping will be disabled and If we put a cross/close icon on right corner

We don't "put a cross/close icon on right corner", that's the flutter standard for pages with `fullscreenDialog: true`, e.g.:
```dart
      Navigator.push<String>(
        context,
        MaterialPageRoute<String>(
          builder: (BuildContext context) => CropPage(File(inputPath)),
          fullscreenDialog: true, // here
        ),
      );
```

> It will disable all controls to crop/click new image and enable swiping back.
> Though, Haven't considered how we are going to implement it ,But would it serve the purpose, If we ever manage to do it.

Understood. So basically you discard #3332 and replace the existing code with something you don't know how to implement (I don't either). For _only_ 4 images, remember.

Anyway, it's a matter of UX choice (then code maintenance), and we could go on arguing which is the best (?) solution.
I've just mocked a scenario in https://github.com/openfoodfacts/smooth-app/issues/3332#issuecomment-1325473463
Feel free to create a similar scenario, so that everyone has a clearer view.

omkarChend1kar committed:
Merge branch 'develop' into swipeable_product_image_view

omkarChend1kar committed:
Formatting changes

omkarChend1kar committed:
Merge branch 'swipeable_product_image_view' of https://github.com/omkarChend1kar/smooth-app into swipeable_product_image_view

omkarChend1kar committed:
Formatting changes

omkarChend1kar committed:
Merge branch 'swipeable_product_image_view' of https://github.com/omkarChend1kar/smooth-app into swipeable_product_image_view

monsieurtanuki reviewed (commented):
Hi @omkarChend1kar!

Unfortunate timing: we were both working on the same files, I wrote significant changes and merged first.
You've solved the conflicts by removing my changes, which in that particular case was really not appropriate.

From #3329 and #3339 we use a different way to get picture providers. That allows us to be more flexible if the image is in an "in-between" status: e.g. the user changed the image but the image is not uploaded yet and we want to display that image locally, before the upload, before the product refresh.

Please solve the conflicts accordingly. If needed, you can create a new PR from scratch - could be faster and easier.
Sorry about that.

omkarChend1kar committed:
feat: 3129 Allow to swipe between product images on the full screen image view (Conflicts after #3332)

omkarChend1kar committed:
feat: 3129 Allow to swipe between product images on the full screen image view (Formatting Changes)

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_gallery_view.dart:
```diff
@@ -1,17 +1,19 @@
+import 'dart:io';
```
No

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_gallery_view.dart:
```diff
@@ -32,14 +34,11 @@ class _ProductImageGalleryViewState extends State<ProductImageGalleryView> {
-
   late Map<ProductImageData, ImageProvider?> _selectedImages;
-
+  bool _isRefreshed = false;
```
No

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_gallery_view.dart:
```diff
@@ -32,14 +34,11 @@ class _ProductImageGalleryViewState extends State<ProductImageGalleryView> {
   ImageProvider? _provideImage(ProductImageData imageData) =>
-      TransientFile.getImageProvider(imageData, _barcode);
-
+      imageData.imageUrl == null ? null : NetworkImage(imageData.imageUrl!);
```
No

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_gallery_view.dart:
```diff
@@ -80,7 +79,7 @@ class _ProductImageGalleryViewState extends State<ProductImageGalleryView> {
             : null,
         leading: SmoothBackButton(
-          onPressed: () => Navigator.maybePop(context),
+          onPressed: () => Navigator.maybePop(context, _isRefreshed),
```
No

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_gallery_view.dart:
```diff
@@ -119,22 +125,54 @@ class _ProductImageGalleryViewState extends State<ProductImageGalleryView> {
+          },
         ),
       );
+  Future<void> _newImage(ProductImageData data) async {
```
No

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_viewer.dart:
```diff
@@ -3,181 +3,171 @@ import 'dart:io';
-  void dispose() {
-    _localDatabase.upToDate.loseInterest(_barcode);
-    super.dispose();
+    _isImageUrlAvailable = widget.imageData.imageUrl != null;
```
No

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_image_viewer.dart:
```diff
@@ -3,181 +3,171 @@ import 'dart:io';
+  late final AppLocalizations _appLocalizations = AppLocalizations.of(context);
 
-  String get _barcode => _initialProduct.barcode!;
+  /// When the image is edited, this is the new image
```
No

monsieurtanuki reviewed (commented):
Obviously your PR was ready before you received my previous message.
Your merge conflict solving is not correct; please refer to my previous message.

omkarChend1kar commented:
> Hi @omkarChend1kar!
> 
> Unfortunate timing: we were both working on the same files, I wrote significant changes and merged first. You've solved the conflicts by removing my changes, which in that particular case was really not appropriate.
> 
> From #3329 and #3339 we use a different way to get picture providers. That allows us to be more flexible if the image is in an "in-between" status: e.g. the user changed the image but the image is not uploaded yet and we want to display that image locally, before the upload, before the product refresh.
> 
> Please solve the conflicts accordingly. If needed, you can create a new PR from scratch - could be faster and easier. Sorry about that.

Seems like making new branch from `develop` and making changes on top of it for swiping effect would be easier.

monsieurtanuki commented:
@omkarChend1kar Indeed, a refreshed `develop` branch, then a new branch, then a new PR.

omkarChend1kar commented:
> @omkarChend1kar Indeed, a refreshed `develop` branch, then a new branch, then a new PR.

@monsieurtanuki Will I have to close this PR, Before creating new PR for same issue.

monsieurtanuki commented:
@omkarChend1kar I've just converted the current PR to draft, to avoid confusion. We can even close it now (and we will close it anyway).
Just try not to use the exact same title for your new PR ;) - which shouldn't be a problem for github, but probably for us.

omkarChend1kar commented:
> @omkarChend1kar I've just converted the current PR to draft, to avoid confusion. We can even close it now (and we will close it anyway). Just try not to use the exact same title for your new PR ;) - which shouldn't be a problem for github, but probably for us.

Ohkay!!

monsieurtanuki commented:
Closed in favor of #3363.
